File: Classical-Problems/Producer-Consumer/producer-consumer.py
import threading
import time
import random
import sys

buffer = [] # Items buffer (just a list)
buffer_lock = threading.Lock()
item_produced = threading.Condition()

num_consumers = 100
num_producers = 3

def Producer(id):
    global buffer
    global buffer_lock
    global item_produced
    # Producer pseudo (from textbook):
    # event = waitForEvent()
    # buffer.add(event)
    print(f"producer {id} starts")
    for _ in range((num_consumers//3)+1): # wastefully large right now, maybe only produce when it becomes empty?
        added = random.randint(0,1)
        with buffer_lock:
            print(f"producer {id} adds {added} to the buffer")
            buffer.append(added)
        with item_produced:
            print(f"producer {id} notifies consumers they can consume")
            item_produced.notify_all()
            
    print(f"producer {id} exits")
        
    

def Consumer(id):
    global buffer
    global buffer_lock
    global item_produced
    # Consumer pseudo (from textbook):
    # event = buffer.pop()
    # event.process() (do something with event)
    print(f"consumer {id} starts")
    
    with item_produced:
        print(f"consumer {id} starts waiting loop")
        
        
        with buffer_lock:
            while len(buffer) == 0:
                buffer_lock.release()
                print(f"consumer {id} waits for an item to be produced")
                item_produced.wait()
                print(f"consumer {id} returns from waiting")
                buffer_lock.acquire()
            
            item = buffer.pop()
            print(f"consumer {id} reads {item} from buffer")
    
    print(f"consumer {id} exits")
    


if __name__=="__main__":
    for i in range(num_consumers):
        ci = threading.Thread(target=Consumer, args=(i,))
        ci.start()

    for i in range(num_producers):
        pi = threading.Thread(target=Producer, args=(i,))
        pi.start()
            
    # Cleanup
    parent = threading.current_thread()
    for thread in threading.enumerate():
        if thread is not parent:
            thread.join()

File: Classical-Problems/Producer-Consumer/producer-consumer.rs
use std::thread;
use std::time::Duration;
use std::sync::Condvar;
use std::sync::Mutex;
use std::sync::MutexGuard;
use std::sync::Arc;

const NUM_PRODUCERS: i32 = 3;
const NUM_CONSUMERS: i32 = 100;

fn main() {
    // Track number of threads
    let mut threads = vec![];

    let triple = Arc::new((Mutex::new(false), Mutex::new(vec![0]), Condvar::new()));

    for i in 0..NUM_CONSUMERS {
        let c_triple = Arc::clone(&triple);
        // Consumers spawner
        threads.push(thread::spawn(move || {
            // Consumer code
            println!("consumer {} starts", i);

            let (buff_empty_lock, buff_lock, item_produced) = &*c_triple;

            let mut buff_empty = buff_empty_lock.lock().unwrap();

            while *buff_empty { 
                println!("consumer {} waits for buffer to not be empty",i);
                buff_empty = item_produced.wait(buff_empty).unwrap();
            }

            // Now we should know the buffer is NOT empty

            // Consume and print
            let item = buff_lock.lock().unwrap().pop();
            *buff_empty = buff_lock.lock().unwrap().len() == 0;

            println!("consumer {} consumed {:?} from buffer",i,item);

            println!("consumer {} exits", i);
        }));

        // thread::sleep(Duration::from_secs(1));
    }

    for j in 0..NUM_PRODUCERS {
        let c_triple = Arc::clone(&triple);
        // Producer spawner
        threads.push(thread::spawn(move || {
            // Producer code
            println!("producer {} starts", j);

            let (buff_empty_lock, buff_lock, item_produced) = &*c_triple;

            let mut buff_empty = buff_empty_lock.lock().unwrap();

            for _ in 0..NUM_CONSUMERS/NUM_PRODUCERS+1 {
                // Wastefully large prodduction right now
                println!("producer {} produces 0 to the buffer",j);
                buff_lock.lock().unwrap().push(0);

                *buff_empty = false;
                println!("producer {} notifies all consumers",j);
                item_produced.notify_all();
            }

            println!("producer {} exits", j);
        }));

        // thread::sleep(Duration::from_secs(1));
    }

    for t in threads {
        let _ = t.join().unwrap();
    }
}